packages/nextjs/utils/auth.ts

```diff
@@ -0,0 +1,127 @@
+import { cookies } from "next/headers";
+import { AuthOptions, JWT, Session, User, getServerSession } from "next-auth";
+import CredentialsProvider from "next-auth/providers/credentials";
+import { getCsrfToken } from "next-auth/react";
+import { createPublicClient, http } from "viem";
+import { mainnet } from "viem/chains";
+import { type SiweMessage, parseSiweMessage, validateSiweMessage } from "viem/siwe";
+import { UserRole } from "~~/services/database/config/types";
+import { getUserByAddress } from "~~/services/database/repositories/users";
+
+export const publicClient = createPublicClient({
+  chain: mainnet,
+  transport: http(),
+});
+
+export const providers = [
+  CredentialsProvider({
+    name: "Ethereum",
+    credentials: {
+      message: {
+        label: "Message",
+        type: "text",
+        placeholder: "0x0",
+      },
+      signature: {
+        label: "Signature",
+        type: "text",
+        placeholder: "0x0",
+      },
+    },
+    async authorize(credentials) {
+      try {
+        if (!credentials?.message || !credentials?.signature) {
+          return null;
+        }
+
+        const siweMessage = parseSiweMessage(credentials.message) as SiweMessage;
+
+        const isMessageValid = validateSiweMessage({
+          address: siweMessage?.address,
+          message: siweMessage,
+        });
+
+        if (!isMessageValid) {
+          return null;
+        }
+
+        const nextAuthUrl =
+          new URL(process.env.NEXTAUTH_URL as string) ||
+          (process.env.VERCEL_URL ? `https://${process.env.VERCEL_URL}` : null);
+
+        if (!nextAuthUrl || siweMessage.domain !== nextAuthUrl.host) {
+          return null;
+        }
+
+        const csrfToken = await getCsrfToken({
+          req: {
+            headers: {
+              cookie: cookies().toString(),
+            },
+          },
+        });
+
+        if (siweMessage.nonce !== csrfToken) {
+          return null;
+        }
+
+        const isSignatureValid = await publicClient.verifyMessage({
+          address: siweMessage.address,
+          message: credentials.message,
+          signature: credentials.signature as `0x${string}`,
+        });
+
+        if (!isSignatureValid) {
+          return null;
+        }
+
+        const user = await getUserByAddress(siweMessage.address);
+
+        if (!user) {
+          return null;
+        }
+
+        return { id: user.userAddress, role: user.role };
```

Comment:
NOTE: Here we don't create user unlike other apps. We just set the session if the user is present and if not we don't set the session at all. 

Since this is for just admin, we preassume they are already present in DB